### ISSUE TITLE
[Bug]: Automatic respawn & some respawn offset

### ISSUE THREAD

**Issue by Looodon** [someaddons/respawningstructures#10]
### Describe the bug you're experiencing

Bug 1: I tried the lastest version of the mod (3.0) with no other mods installed, but setting "minutesUntilRespawn" to any number still doesn't work for me, in the log I had it set to 1 minute. I did the test with "blockCountMod" set to 0.0 and another time set to 1.0. 

Bug 2: when I manually respawned shipwrecks, they spawned above the naturally generated ones. This doesn't happen to all structures. This issue happened with some modded structures like mowziesmobs:umvuthana_grove for example. 
![Screen Shot (2024-12-12 20_59_43)](https://github.com/user-attachments/assets/9b73fe74-0786-4ae7-a98c-b26e1f77b987)

### Feedback
I love the config change to include "blockCountMod" & "respawnableStructures", I didn't get to testing the whitelist after seeing the autorespawn didn't work. Looking forward to using it once this is fixed, thank you for working on the mod, I love the idea

### Reproducability

1. Create new modpack, add the mod
2. Launch the game, generate the config, close the game
3. set "minutesUntilRespawn": to any number 
4. break a structure (mansion for example) and wait

### Mod up to date

- [X] Before submitting this issue I updated to the newest version and reproduced it

### Minecraft version

1.20.1

### Modloader version

Forge

### Logs

https://pastebin.com/LVsYX4yy

**Comment by someaddons** [comment]
that is because breaking just a couple blocks is not enough usage to trigger a respawn^^
Try opening a chest in addition

**Comment by Looodon** [comment]
After more testing, It seems like TNT doesn't get registered as blocks broken so the whole structure could be exploded & it doesn't respawn. I broke blocks by hand and had to in some cases it respawned, but not all cases. 

There are instances where some structures will be broken and not looted or where mods introduce boss liars where it would be nice to be able to fight them multiple times. Some structure don't have chests.

Would it be possible to add a config to make structures respawn no matter what?
Config idea 1: guaranteed respawn no matter what, even with no interaction
Config idea 2: guaranteed respawn if 1 mob is killed
Config idea 3: guaranteed respawn if 1 block is broken or placed
This would make the mod flexible, avoid any inconsistencies, & solve the issue. 

There's also a bug where the structure spawns like 3 blocks higher than it's supposed to (shipwrecks, etc)

EDIT: I just tested the whitelist and it didn't work, did I do something wrong?
![Screen Shot (2024-12-13 03_26_40)](https://github.com/user-attachments/assets/2ae1045b-0d47-42e2-96d3-729c02d2dc75)

**Comment by someaddons** [comment]
seems like the whitelist doesnt work as intended, the "respawnablestructures" however should force the respawn, regardless of containers looted, but in this case is blocked by the blacklist.

>There's also a bug where the structure spawns like 3 blocks higher than it's supposed to (shipwrecks, etc)

That is not an actual bug, as my mod just tells the structure to regenerate, some structures simply place themselves like that over existing blocks

**Comment by someaddons** [comment]
Fixed a couple things in the next version, spawn positions should be better for some structures aswell
3.1:
Fix whitelist
Bossmobs now should trigger a respawn when killed
Count explosions within structures as blocksbroken
Improve respawn position for Template structures

**Comment by Looodon** [comment]
Hey! 

I'm very grateful you took the time to work on the mod! The whitelist now works! Boss kill counts as "22 mobs killed", and TNT counts as blocks destroyed. The respawn position still happens sometimes, but not always.

According to my test, boss kill alone doesn't respawn the structure, you have to loot chests (some structures don't have chests). The video below shows one of the tests where killing the boss didn't respawn the structure (it respawned after I looted chests after the vid). It would be nice if boss kill alone was enough to respawn it, I'll make a few more tests, let me know if you need me to test something specific

https://github.com/user-attachments/assets/7b934b2d-b1b0-4556-9e08-0027debdb91d

**Comment by someaddons** [comment]
hm the boss kill actually should be enough, it counts as 10 mob kills and the
"usage threshold" does this:
        if ((spawnerActivations * 3 + containerLooted * 10 + lightsPlaced * 3 + blocksPlaced + blocksBroken + mobsKilled * 4 + playerDeaths * 10) > 30)
        {
            return RespawnStatus.PENDING_RESPAWN;
        }
        
Note that respawning may take like 5-6mins minimum, timers are only updated every 5mins

**Comment by Looodon** [comment]
I just ran another test, I killed the umvuthana boss from mowzies mobs without breaking blocks/looting I had:
"respawn status: UNUSED, Remaining minutes until respawn: 1"
Then it changed to:
"respawn status: UNUSED, Remaining minutes until respawn: 0"
It stayed like that for 10 mins, then I broke 1 block and it changed to: 
"respawn status: WAITING_RESPAWN_TIME, Remaining minutes until respawn: 1"
It stayed like that for 5 mins then the structure respawned

**Comment by someaddons** [comment]
funny, I cannot reproduce that :D
btw a "boss" is just a mob with > 80 hp
to test I killed 10 mobs inside the structure, which did switch to pending respawn from unused after a while^^

Did it show 10+ mobs killed after killing the boss?

**Comment by Looodon** [comment]
Ooh, maybe the issue is coming from the fact is a modded mob? The boss has 150 hearts, it counts as 22 mob kills when I defeat it. This is what happens, it only triggers when i break a block:
https://github.com/user-attachments/assets/8b570937-b567-43d0-9494-fa463bad7aee

**Comment by someaddons** [comment]
quite odd :D also that it counts for 22, I only made 1 or 10 for bosses,
guess I'll try with that mowzies mobs boss

**Comment by someaddons** [comment]
ok it does work for me with mowzies boss aswell, says 22 after killed and respawned after roughly 5 minutes :D
did you put any ID's in the config's black/whitelist/respawning lists?

**Comment by Looodon** [comment]
sorry I had to go to bed & then editing the video took a while, I recorded a video of me creating a fresh new modpack, downloading only mowzies & respawning mod. The first video shows how the "respawn status" stays on "unused" after killing, breaking blocks, and looting. I didn't wait for 5 minutes, it was just to show the "respawn status" with the original config:
https://github.com/user-attachments/assets/e01a3553-654f-4fe8-bfc1-64184da815e6

In the second video, I record for 10 minutes, but it's very sped up. With 30 broken blocks, 1 container looted, and 22 mob kills I wait around 8 minutes with no action until i use TNT changing it to 968 blocks broken, then "respawn status" changes to "waiting respawn time" (at 9:34 timer). I use my custom config which is this [config](https://pastebin.com/FxYquGPQ).
https://github.com/user-attachments/assets/fdffbc23-7489-4992-8fc3-79c18818af85

**Comment by someaddons** [comment]
Pretty sure I figured it out, you're using a new world to test right? The new world needs 5 minutes to initialize its own timer, before that timer has useable info for the structures being modified. So pretty much gotta wait at least 5 minutes before doing anything, then killing the boss should no longer result in UNUSED status^^

**Comment by Looodon** [comment]
YES! I think that was the issue, I just did it on a world older than 5 minutes and it worked as intended. The structures respawning offset higher issue is better than before & it doesn't duplicate chests so I dont think it's a big deal. I'm so happy with the whitelist addition, boss defeat triggering respawn, & respawnableStructures always respawn. I think the mod is in a perfect spot & I cant wait to use it. Thank you so much for making it & taking the time to improve it, I consider this solved! :D

**Comment by someaddons** [comment]
nice :D
I did do some more work though:

3.2:
Add feature which tracks player respawn positions and prevents nearby structure respawns
Add feature which prevents respawn in too busy chunks, aka chunks which have been loaded a lot during the respawn timer
Detects blockentity placement and keeps track of the amount in a structure, which can also prevent a respawn(player activity indication)

Btw the "playerNearbyTime" config will break your testing, since it checks how much % of the respawn time a player was nearby, might want to disable it when testing

**Comment by Looodon** [comment]
So if i understand correctly:
> Add feature which tracks player respawn positions and prevents nearby structure respawns

^This is sees if you respawn nearby a structure and stops it from respawning, an example is setting a bed in the structure or outside

> Add feature which prevents respawn in too busy chunks, aka chunks which have been loaded a lot during the respawn timer

^If someone stayed in a chunk for 10 hours, the structure will see that and not respawn (this is to prevent homes being affected)

> Detects blockentity placement and keeps track of the amount in a structure, which can also prevent a respawn(player activity indication)

^This tracks if u placed multiple chests, another way of tracking homes

> "playerNearbyTime" checks how much % of the respawn time a player was nearby

^tracks if a player was nearby when a respawn happened, another way of knowing if someone lives there

In other words, best chance of a respawn is to not live nearby or "camp" the structure/boss. Does "respawnableStructures" bypass all these?

**Comment by someaddons** [comment]
yes "respawnableStructures" skips all those checks
Those checks are mainly to not respawn structures in the player's bases, as they may destroy their stuff there
